File: ROUND_1/CPP/Optimal/Data_fusion.cpp
#include <iostream>
#include <iomanip>
#include <vector>
#include <algorithm>
using namespace std;

// Values are within +/- 10^6, so these act as -infinity and +infinity
const int NEG_INF = -2000000000;
const int POS_INF = 2000000000;

int main() {
    int n;
    cin >> n;

    vector<int> a(n);
    for (int i = 0; i < n; i++)
        cin >> a[i];

    int m;
    cin >> m;

    vector<int> b(m);
    for (int i = 0; i < m; i++)
        cin >> b[i];

    // Always binary search on the smaller array
    if (n > m) {
        swap(a, b);
        swap(n, m);
    }

    int low = 0, high = n;

    while (low <= high) {
        int cutA = (low + high) / 2;
        int cutB = (n + m + 1) / 2 - cutA;

        int leftA = (cutA == 0) ? NEG_INF : a[cutA - 1];
        int rightA = (cutA == n) ? POS_INF : a[cutA];

        int leftB = (cutB == 0) ? NEG_INF : b[cutB - 1];
        int rightB = (cutB == m) ? POS_INF : b[cutB];

        if (leftA <= rightB && leftB <= rightA) {

            // Odd total: median is the largest element on the left
            if ((n + m) % 2 == 1) {
                cout << max(leftA, leftB) << endl;
                return 0;
            }

            double median = (max(leftA, leftB) + min(rightA, rightB)) / 2.0;

            if (median == (int)median)
                cout << (int)median << endl;
            else
                cout << fixed << setprecision(1) << median << endl;

            return 0;
        }

        if (leftA > rightB)
            high = cutA - 1;
        else
            low = cutA + 1;
    }

    return 0;
}
/*
Time Complexity: O(log(min(N, M)))
Space Complexity: O(1)
*/

File: ROUND_1/JAVA/Optimal/Data_fusion.java

import java.io.*;
import java.util.*;


public class Data_fusion {
     public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        int n = sc.nextInt();
        int[] nums1 = new int[n];
        for (int i = 0; i < n; i++)
            nums1[i] = sc.nextInt();

        int m = sc.nextInt();
        int[] nums2 = new int[m];
        for (int i = 0; i < m; i++)
            nums2[i] = sc.nextInt();

        // Always binary search on the smaller array
        if (n > m) {
            int[] temp = nums1;
            nums1 = nums2;
            nums2 = temp;

            int t = n;
            n = m;
            m = t;
        }

        int low = 0;
        int high = n;

        while (low <= high) {

            int partitionX = (low + high) / 2;
            int partitionY = (n + m + 1) / 2 - partitionX;

            int maxLeftX = (partitionX == 0)
                    ? Integer.MIN_VALUE
                    : nums1[partitionX - 1];

            int minRightX = (partitionX == n)
                    ? Integer.MAX_VALUE
                    : nums1[partitionX];

            int maxLeftY = (partitionY == 0)
                    ? Integer.MIN_VALUE
                    : nums2[partitionY - 1];

            int minRightY = (partitionY == m)
                    ? Integer.MAX_VALUE
                    : nums2[partitionY];

            if (maxLeftX <= minRightY && maxLeftY <= minRightX) {

                if ((n + m) % 2 == 0) {
                    double median = (Math.max(maxLeftX, maxLeftY)
                                   + Math.min(minRightX, minRightY)) / 2.0;

                    if (median == (int) median)
                        System.out.println((int) median);
                    else
                        System.out.printf("%.1f\n", median);

                } else {
                    System.out.println(Math.max(maxLeftX, maxLeftY));
                }

                return;
            }

            else if (maxLeftX > minRightY) {
                high = partitionX - 1;
            }

            else {
                low = partitionX + 1;
            }
        }

        sc.close();
    }
    
}
/* 
Time Complexity
O(log(min(N, M)))
Space Complexity
O(1)
*/
